tests/serialize.rs

```diff
@@ -0,0 +1,386 @@
+            "input_to_tracked": {
+              "0:3": {
+                "value": {
+                  "index": 3073,
+                  "generation": 0
+                },
+                "verified_at": {
+                  "data": 1
+                },
+                "revisions": {
+                  "changed_at": 1,
+                  "durability": "Low",
+                  "origin": {
+                    "Derived": [
+                      {
+                        "key": {
+                          "key_index": {
+                            "index": 3,
+                            "generation": 0
+                          },
+                          "ingredient_index": 1
+                        }
+                      },
+                      {
+                        "key": {
+                          "key_index": {
+                            "index": 3073,
+                            "generation": 0
+                          },
+                          "ingredient_index": 2147483651
+                        }
+                      }
+                    ]
+                  },
+                  "verified_final": true,
+                  "extra": {
+                    "tracked_struct_ids": [
+                      [
+                        {
+                          "ingredient_index": 3,
+                          "hash": 6073466998405137972,
+                          "disambiguator": 0
+                        },
+                        {
+                          "index": 3073,
+                          "generation": 0
+                        }
+                      ]
+                    ],
+                    "cycle_heads": [],
+                    "iteration": 0
+                  }
+                }
+              }
+            },
+            "unit_to_interned": {
+              "6:1025": {
+                "value": {
+                  "index": 2049,
+                  "generation": 0
+                },
+                "verified_at": {
+                  "data": 1
+                },
+                "revisions": {
+                  "changed_at": 1,
+                  "durability": "High",
+                  "origin": {
+                    "Derived": [
+                      {
+                        "key": {
+                          "key_index": {
+                            "index": 2049,
+                            "generation": 0
+                          },
+                          "ingredient_index": 2
+                        }
+                      }
+                    ]
+                  },
+                  "verified_final": true,
+                  "extra": null
+                }
+              }
+            }
+          }
+        }"#]];
+
+    expected.assert_eq(&serialized);
+
+    let mut db = common::EventLoggerDatabase::default();
+    <dyn salsa::Database>::deserialize(
+        &mut db,
+        &mut serde_json::Deserializer::from_str(&serialized),
+    )
```

Comment:
Does `deserialize` bump the revision to `2` to force a validation of all queries?

Reply:
Hmm no, should it?

Reply:
I think there are a few things to consider here

* memos and other ingredients that weren't verified in the last revision (current when deserializing) still need to be dirty. The easiest here might be to only serialize ingredients that are verified in the current revision and we revisit this later
* we need to bump the revision if any input changes during deserialization. Now, we don't support this yet so it's fine to ignore for now.